File: examples/go/core/api.case.go
package core

import "app-protocol/examples/go/core/shared"

type ApiContext struct {
	shared.AppBaseContext
	HTTPClient shared.AppHttpClient
	DB         any
	Auth       any
	Storage    shared.AppStorageClient
	Cache      shared.AppCache
	Cases      shared.Dict
	Packages   shared.Dict
	Extra      shared.Dict
}

type ApiResponse[T any] struct {
	shared.AppResult[T]
	StatusCode int `json:"statusCode,omitempty"`
}

func (r ApiResponse[T]) HTTPStatus() int {
	return r.StatusCode
}

type RouteRequest struct {
	Body       any
	Method     string
	Path       string
	Params     map[string]string
	RawRequest any
}

type RouteBinding struct {
	Method  string
	Path    string
	Handler func(request RouteRequest) (any, error)
}

type APIHandler interface {
	HandleAny(input any) (any, error)
}

func ExecuteAPI[TInput, TOutput any](
	input TInput,
	validate func(TInput) error,
	authorize func(TInput) error,
	composition func(TInput) (TOutput, error),
	service func(TInput) (TOutput, error),
) ApiResponse[TOutput] {
	if validate != nil {
		if err := validate(input); err != nil {
			appErr := shared.ToAppError(err, "INTERNAL")
			return ApiResponse[TOutput]{
				AppResult: shared.AppResult[TOutput]{
					Success: false,
					Error:   &appErr,
				},
			}
		}
	}

	if authorize != nil {
		if err := authorize(input); err != nil {
			appErr := shared.ToAppError(err, "INTERNAL")
			return ApiResponse[TOutput]{
				AppResult: shared.AppResult[TOutput]{
					Success: false,
					Error:   &appErr,
				},
			}
		}
	}

	var (
		result TOutput
		err    error
	)

	switch {
	case composition != nil:
		result, err = composition(input)
	case service != nil:
		result, err = service(input)
	default:
		err = &shared.AppCaseError{
			Code:    "INTERNAL",
			Message: "BaseApiCase: at least one of _service or _composition must be implemented",
		}
	}

	if err != nil {
		appErr := shared.ToAppError(err, "INTERNAL")
		return ApiResponse[TOutput]{
			AppResult: shared.AppResult[TOutput]{
				Success: false,
				Error:   &appErr,
			},
		}
	}

	return ApiResponse[TOutput]{
		AppResult: shared.AppResult[TOutput]{
			Success: true,
			Data:    &result,
		},
	}
}

File: examples/kotlin/core/api.case.kt
package app.protocol.examples.kotlin.core.api

import app.protocol.examples.kotlin.core.Dict
import app.protocol.examples.kotlin.core.shared.AppBaseContext
import app.protocol.examples.kotlin.core.shared.AppCache
import app.protocol.examples.kotlin.core.shared.AppCaseError
import app.protocol.examples.kotlin.core.shared.AppError
import app.protocol.examples.kotlin.core.shared.AppHttpClient
import app.protocol.examples.kotlin.core.shared.AppStorageClient

interface ApiContext : AppBaseContext {
    val httpClient: AppHttpClient?
    val db: Any?
    val auth: Any?
    val storage: AppStorageClient?
    val cache: AppCache?
    val cases: Dict<Any?>?
    val packages: Dict<Any?>?
    val extra: Dict<Any?>?
}

data class ApiResponse<T>(
    val success: Boolean,
    val data: T? = null,
    val error: AppError? = null,
    val statusCode: Int? = null,
)

data class RouteBinding(
    val method: String,
    val path: String,
    val handler: (suspend (RouteRequest) -> Any?)? = null,
)

data class RouteRequest(
    val body: Any? = null,
    val method: String,
    val path: String,
    val params: Map<String, String> = emptyMap(),
    val request: Any? = null,
)

abstract class BaseApiCase<TInput, TOutput>(
    protected val ctx: ApiContext,
) {
    abstract suspend fun handler(input: TInput): ApiResponse<TOutput>

    open fun router(): RouteBinding? = null

    open suspend fun test() {
    }

    protected open suspend fun _validate(input: TInput) {
    }

    protected open suspend fun _authorize(input: TInput) {
    }

    protected open fun _repository(): Any? = null

    protected open val usesService: Boolean = false

    protected open val usesComposition: Boolean = false

    protected open suspend fun _service(input: TInput): TOutput? = null

    protected open suspend fun _composition(input: TInput): TOutput? = null

    protected suspend fun execute(input: TInput): ApiResponse<TOutput> =
        try {
            _validate(input)
            _authorize(input)

            if (!usesService && !usesComposition) {
                throw AppCaseError(
                    code = "INTERNAL",
                    message = "BaseApiCase: at least one of _service or _composition must be implemented",
                )
            }

            val result = when {
                usesComposition -> _composition(input)
                usesService -> _service(input)
                else -> null
            }

            if (result == null) {
                throw AppCaseError(
                    code = "INTERNAL",
                    message = "BaseApiCase: execution center returned null",
                )
            }

            ApiResponse(
                success = true,
                data = result,
            )
        } catch (error: AppCaseError) {
            ApiResponse(
                success = false,
                error = error.toAppError(),
            )
        }
}
